Ffeldberg/rel 8140/add view features to tf provider

monsagri opened the pull request:
# Add LaunchDarkly Views Resource Support 

Adds 
* View Resource
* View Linkage Resource
* Data sources for above
* Tests
* Examples

monsagri commented:
I've used a tf file to use all the new sources/resources in our waterbear tf test account, see [here](https://app.launchdarkly.com/projects/views-test/)

```
terraform {
  required_providers {
    launchdarkly = {
      source  = "launchdarkly/launchdarkly"
      version = "~> 2.0"
    }
  }
}

locals {
  # Maintainer ID is needed if using a service token
  # Staging maintainer ID
  # maintainer_id = "68501f4cb6d802092797c654"
  # Staging team member IDs
  # team_member_ids = ["68501f4cb6d802092797c654"]
}

# Configure the LaunchDarkly provider
provider "launchdarkly" {
  # Staging overrides
  # access_token = "api-ffa33721-6c3b-46c3-b0ad-1e16f821dc65"
  # api_host      = "https://ld-stg.launchdarkly.com"
}

# Create a simple project for testing views
resource "launchdarkly_project" "test" {
  name = "Views Test Project"
  key  = "views-test"
  environments {
    name = "development"
    key  = "dev"
    color = "00ff00"
  }
  environments {
    name = "production"
    key  = "prod"
    color = "ff0000"
  }
}

# Create 30 feature flags for testing views
resource "launchdarkly_feature_flag" "backend_flags" {
  count          = 30
  project_key    = launchdarkly_project.test.key
  key            = "backend-flag-${count.index + 1}"
  name           = "Backend Flag ${count.index + 1}"
  description    = "Flag for backend team - will be linked to backend view"
  variation_type = "boolean"
}

resource "launchdarkly_feature_flag" "flag_frontend_1" {
  project_key    = launchdarkly_project.test.key
  key            = "frontend-flag-1"
  name           = "Frontend Flag 1"
  description    = "Flag for frontend team - will be linked to frontend view"
  variation_type = "boolean"
}

resource "launchdarkly_feature_flag" "flag_random_1" {
  project_key    = launchdarkly_project.test.key
  key            = "random-flag-1"
  name           = "Random Flag 1"
  description    = "Flag for no team - will be linked to no view"
  variation_type = "boolean"
}

# Create views for testing
resource "launchdarkly_view" "backend_view" {
  name        = "Backend Team View"
  key         = "backend-view"
  description = "View for backend team"
  project_key = launchdarkly_project.test.key
  tags = ["backend", "api"]
  # maintainer_id = local.maintainer_id
}

resource "launchdarkly_view" "frontend_view" {
  name        = "Frontend Team View"
  key         = "frontend-view"
  description = "View for frontend team"
  project_key = launchdarkly_project.test.key
  tags = ["frontend", "ui"]
  # maintainer_id = local.maintainer_id
}


resource "launchdarkly_view" "no_team_view" {
  name        = "No Team View"
  key         = "no-team-view"
  description = "View for no team"
  project_key = launchdarkly_project.test.key
  tags = ["frontend", "ui"]
  # maintainer_id = local.maintainer_id
}

# Link flags to views
resource "launchdarkly_view_links" "backend_view_links" {
  view_key    = launchdarkly_view.backend_view.key
  project_key = launchdarkly_project.test.key
  
  flags = [for flag in launchdarkly_feature_flag.backend_flags : flag.key]
  
  comment = "Linking all 30 backend flags to backend view"
}

resource "launchdarkly_view_links" "frontend_view_links" {
  view_key    = launchdarkly_view.frontend_view.key
  project_key = launchdarkly_project.test.key
  
  flags = [
    launchdarkly_feature_flag.flag_frontend_1.key
  ]
  
  comment = "Linking frontend flags to frontend view"
}

resource "launchdarkly_view_links" "no_team_view_links" {
  view_key    = launchdarkly_view.no_team_view.key
  project_key = launchdarkly_project.test.key
  
  flags = [
  ]
  
  comment = "Not linking any flags to no team view"
}

# Create a custom role that grants access to flags in the backend-view
resource "launchdarkly_custom_role" "backend_view_access" {
  key         = "backend-view-access"
  name        = "Backend View Access"
  description = "Grants access to all flags contained in the backend-view"
  
  policy_statements {
    resources = [
      "proj/*:env/*:flag/*;view:backend-view"
    ]
    actions = ["*"]
    effect  = "allow"
  }
}

# Create a team with the custom role
resource "launchdarkly_team" "backend_team" {
  key         = "backend-team"
  name        = "Backend Development Team"
  description = "Team for backend developers with access to backend view flags"
  
  custom_role_keys = [
    launchdarkly_custom_role.backend_view_access.key
  ]
  
  # Add some member IDs if you have them available
  # member_ids = locals.team_member_ids
}

# Output useful information for verification
output "project_info" {
  value = {
    id   = launchdarkly_project.test.id
    key  = launchdarkly_project.test.key
    name = launchdarkly_project.test.name
  }
}
```

sloloris commented on launchdarkly/resource_launchdarkly_view.go:
```diff
@@ -0,0 +1,245 @@
+			MAINTAINER_ID: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The member ID of the maintainer for this view.",
```
```suggestion
				Description:   "The member ID of the maintainer for this view. Either this or `maintainer_team_key` must be set.",
```

sloloris commented on launchdarkly/resource_launchdarkly_view.go:
```diff
@@ -0,0 +1,245 @@
+			MAINTAINER_TEAM_KEY: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The team key of the maintainer team for this view.",
```
```suggestion
				Description:   "The team key of the maintainer team for this view. One of `maintainer_id` and `maintainer_team_key` must be set.",
```

sloloris commented on launchdarkly/resource_launchdarkly_view.go:
```diff
@@ -0,0 +1,245 @@
+		viewPost["generateSdkKeys"] = generateSdkKeys.(bool)
+	}
+
+	if maintainerId, ok := d.GetOk(MAINTAINER_ID); ok {
+		viewPost["maintainerId"] = maintainerId.(string)
+	}
+
+	if maintainerTeamKey, ok := d.GetOk(MAINTAINER_TEAM_KEY); ok {
+		viewPost["maintainerTeamKey"] = maintainerTeamKey.(string)
+	}
```
you can but don't have to enforce at least one or the other of these but not both here - the API will fail if not

sloloris replied:
nvm view_helper.go wasn't expanded

sloloris replied:
okay actually sorry on second thought i think you should enforce these, because of the way we calculate it in gonfalon. if you define neither, the API will still return a maintainer_id, which i'm fairly certain will cause terraform to get confused about the state. there's no way to make maintainer_id optional but computed only when team_maintainer isn't defined, so i think the best way to do this is to hard enforce one or the other on the terraform side

sloloris commented on launchdarkly/resource_launchdarkly_view_links.go:
```diff
@@ -0,0 +1,261 @@
+			// 		ValidateDiagFunc: validateKey(),
+			// 	},
+			// },
+			"comment": {
```
can we just stick this one in keys.go too?

sloloris replied:
on second thought i think we should probably omit this from the terraform resource, it doesn't really make a lot of sense since the resources don't correlate to actions but rather states

sloloris commented on launchdarkly/resource_launchdarkly_view_test.go:
```diff
@@ -0,0 +1,249 @@
+				Config: fmt.Sprintf(testAccViewCreate, projectKey, viewKey, viewName, viewDescription),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckViewExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, NAME, viewName),
+					resource.TestCheckResourceAttr(resourceName, DESCRIPTION, viewDescription),
+					resource.TestCheckResourceAttr(resourceName, GENERATE_SDK_KEYS, "false"),
+					resource.TestCheckResourceAttr(resourceName, "tags.#", "1"),
```
```suggestion
					resource.TestCheckResourceAttr(resourceName, PROJECT_KEY, projectKey),
					resource.TestCheckResourceAttr(resourceName, KEY, viewKey),
					resource.TestCheckResourceAttr(resourceName, NAME, viewName),
					resource.TestCheckResourceAttr(resourceName, DESCRIPTION, viewDescription),
					resource.TestCheckResourceAttr(resourceName, GENERATE_SDK_KEYS, "false"),
					resource.TestCheckResourceAttr(resourceName, ARCHIVED, "false"),
					resource.TestCheckResourceAttr(resourceName, "tags.#", "1"),
```

sloloris commented on launchdarkly/resource_launchdarkly_view_links_test.go:
```diff
@@ -0,0 +1,223 @@
+			},
+			{
+				Config: fmt.Sprintf(testAccViewLinksUpdate, projectName, projectKey),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckViewLinksExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "flags.#", "2"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "flags.*", "test-flag-1"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "flags.*", "test-flag-3"),
+					resource.TestCheckResourceAttr(resourceName, "comment", "Updated test linking flags to view"),
+				),
```
can you check at the end (do a defer helper function up top) that these flags were unlinked after these were deleted? i'm not actually sure when the teardown function gets called so you may have to add another step with the flag link resource config removed

monsagri replied:
Yep, good shout!

sloloris reviewed (changes requested):
okay a few main changes to request:
- we should get rid of the comment attribute on the view flags link. it's not really conventional to do comments on terraform stuff because it corresponds to state rather than action
- we should enforce the maintainer at the terraform level. see comment on code
- a few test suggestions

otherwise it looks pretty good. such a pain to build it out for a beta API 😭

mmrj commented on docs/data-sources/feature_flag.md:
````diff
@@ -16,9 +16,49 @@ This data source allows you to retrieve feature flag information from your Launc
 ## Example Usage
 
 ```terraform
+# Example: Get feature flag information including view memberships
````
nit, optional to address: I'm not sure we use "memberships" elsewhere (yet) to describe this. Does this mean the views that a flag is linked to? Maybe something like this?

```suggestion
# Example: Get feature flag information including linked views
```

mmrj commented on docs/data-sources/feature_flag.md:
```diff
@@ -16,9 +16,49 @@ This data source allows you to retrieve feature flag information from your Launc
+  key         = "example-flag"
+}
+
+# The feature flag data source now includes discovery of view memberships
```
```suggestion
# The feature flag data source now includes discovery of linked views
```
(only if you also choose to use the previous suggestion)

mmrj commented on docs/data-sources/feature_flag.md:
```diff
@@ -16,9 +16,49 @@ This data source allows you to retrieve feature flag information from your Launc
+  is_shared_flag         = length(data.launchdarkly_feature_flag.example.views) > 1
+}
+
+# Example: Conditional resource creation based on view membership
```
```suggestion
# Example: Conditional resource creation based on views
```
optional

mmrj commented on docs/data-sources/feature_flag.md:
```diff
@@ -16,9 +16,49 @@ This data source allows you to retrieve feature flag information from your Launc
+  on      = true
+}
+
+# Example: Generate team notifications based on flag view assignments
```
```suggestion
# Example: Generate team notifications based on views
```
optional

mmrj commented on docs/resources/view_links.md:
```diff
@@ -0,0 +1,113 @@
+subcategory: ""
+description: |-
+  Provides a LaunchDarkly view links resource for managing bulk resource linkage to views.
+  This resource allows you to efficiently link multiple flags (and in the future, segments and metrics) to a specific view. This is particularly useful for administrators organizing resources by team or deployment unit.
```
double-checking: do we want to communicate now that views will eventually include segments and metrics?

monsagri replied:
So the current approach is that we won't communicate ahead of time, but we will enable new resources as and when they become available, good callout.

mmrj commented on docs/resources/view_links.md:
```diff
@@ -0,0 +1,113 @@
+    "feature-database-migration",
+    "feature-cache-optimization",
+    "feature-api-versioning",
+    # To add a new flag, simply add it to this list
```
```suggestion
    # To add a new flag, add it to this list
```
style guide recommendation -- we try not to use "simply" and "just" because different tasks have different levels of difficulty for different folks

monsagri replied:
Thanks for the heads up!

mmrj reviewed (approved):
✅ from Docs for the `/docs/*.md` updates, with a few optional suggestions (some of which are suggested in the wrong place, because I'm not familiar enough with Terraform to know where the sources for various generated files are)

monsagri commented on launchdarkly/resource_launchdarkly_view_links.go:
```diff
@@ -0,0 +1,261 @@
+			FLAGS: {
+				Type:        schema.TypeSet,
+				Set:         schema.HashString,
+				Optional:    true,
```
Down the line it'll have to be optional (once we add other resource types, but for now there really is no point in it being optional)

monsagri commented on launchdarkly/resource_launchdarkly_view_test.go:
```diff
@@ -0,0 +1,249 @@
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckViewExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, MAINTAINER_ID, maintainerId),
+				),
```
Done!

monsagri commented:
@sloloris I've just re-requested review, I think I've addressed all of your feedback and added more tests (good shout!)

I left one comment unresolved, the deletion of flag links is working as expected right now, is there something that I'm missing?

sloloris commented on launchdarkly/resource_launchdarkly_view_links.go:
```diff
@@ -0,0 +1,261 @@
+	comment := d.Get("comment").(string)
+
+	// Unlink all flags
+	if flagsRaw, ok := d.GetOk(FLAGS); ok {
```
sorry - hope this wasn't too much of a wild goose chase

monsagri replied:
All good, just used my test.tf config to double check that it worked since I'd seen it working before.

sloloris commented on launchdarkly/resource_launchdarkly_view_test.go:
```diff
@@ -0,0 +1,335 @@
+	}
+}
+
+resource "launchdarkly_view" "test" {
+	project_key       = launchdarkly_project.test.key
+	key               = "%s"
+	name              = "%s"
+	description       = "%s"
+	generate_sdk_keys = true
+	tags              = ["test", "updated"]
+}
```
i guess maybe it just retains its previous state and it gets set back onto the state because the API still returns it?

sloloris commented on .github/workflows/test.yml:
```diff
@@ -98,6 +98,8 @@ jobs:
           - TestAccTeamMember_UpdateWithCustomRole
           - TestAccTeam_
           - TestAccTeamRoleMapping_
+          - TestAccView
+          - TestAccViewLinks
```
```suggestion
          - TestAccView
```
but also `TestAccView` will run everything prepended by that so you don't need to put ViewLinks separately. we separated them out because when too many were running in parallel sometimes they were timing out
